crates/accelerate/src/synthesis/evolution/mod.rs

```diff
@@ -0,0 +1,431 @@
+}
+
+/// Return the vector of Qiskit's gate corresponding to the given vector
+/// of Rustiq's Clifford gate.
+fn qiskit_clifford_gates(gates: &Vec<CliffordGate>) -> Vec<QiskitGate> {
+    let mut qiskit_gates: Vec<QiskitGate> = Vec::with_capacity(gates.len());
+    for gate in gates {
+        qiskit_gates.push(qiskit_clifford_gate(gate));
+    }
+    qiskit_gates
+}
+
+/// Returns the number of CNOTs.
+fn cnot_count(qgates: &[QiskitGate]) -> usize {
+    qgates.iter().filter(|&gate| gate.2.len() == 2).count()
+}
+
+/// Given the Clifford circuit returned by Rustiq's pauli network synthesis algorithm,
+/// generates a sequence of Qiskit gates that implements this circuit.
+/// If `fix_clifford_method` is `0`, the original circuit is inverted; if `1`, it is
+/// resynthesized using Qiskit; and if `2` it is resynthesized using Rustiq.
+fn synthesize_final_clifford(
+    rcircuit: &CliffordCircuit,
+    resynth_clifford_method: usize,
+) -> Vec<QiskitGate> {
+    match resynth_clifford_method {
+        0 => qiskit_clifford_gates(&rcircuit.gates),
+        1 => {
+            // Qiskit-based resynthesis
+            let qcircuit = qiskit_clifford_gates(&rcircuit.gates);
+            let new_qcircuit = resynthesize_clifford_circuit(rcircuit.nqbits, &qcircuit).unwrap();
+            if cnot_count(&qcircuit) < cnot_count(&new_qcircuit) {
+                qcircuit
+            } else {
+                new_qcircuit
+            }
+        }
+        _ => {
+            // Rustiq-based resynthesis
+            let tableau = IsometryTableau::new(rcircuit.nqbits, 0);
+            let new_rcircuit = isometry_synthesis(&tableau, &Metric::COUNT, 1);
+            if new_rcircuit.cnot_count() < rcircuit.cnot_count() {
+                qiskit_clifford_gates(&new_rcircuit.gates)
+            } else {
+                qiskit_clifford_gates(&rcircuit.gates)
+            }
+        }
+    }
+}
+
+/// Calls Rustiq's pauli network synthesis algorithm and returns the
+/// Qiskit circuit data with Clifford gates and rotations.
+///
+/// # Arguments
+///
+/// * py: a GIL handle, needed to add and negate rotation parameters in Python space.
+/// * num_qubits: total number of qubits.
+/// * pauli_network: pauli network represented in sparse format. It's a list
+///     of triples such as `[("XX", [0, 3], theta), ("ZZ", [0, 1], 0.1)]`.
+/// * optimize_count: if `true`, Rustiq's synthesis algorithms aims to optimize
+///     the 2-qubit gate count; and if `false`, then the 2-qubit depth.
+/// * preserve_order: whether the order of paulis should be preserved, up to
+///     commutativity. If the order is not preserved, the returned circuit will
+///     generally not be equivalent to the given pauli network.
+/// * upto_clifford: if `true`, the final Clifford operator is not synthesized
+///     and the returned circuit will generally not be equivalent to the given
+///     pauli network. In addition, the argument `upto_phase` would be ignored.
```

Comment:
If it's the final Clifford, then one can have both "upto_clifford" and "upto_phase" since the global phase should not affect the final measurement?

Reply:
I don't know if it makes sense to ignore the final Clifford but to keep the global phase. And in any case currently we don't have a way to do that (without the extended Clifford simulator). So the assumption that I tried to document here is that if we ignore the final Clifford, we also automatically ignore the global phase. Anyway, for now this is really an experimental option available to users (I am pretty sure that @ajavadia would like to have it). I also expect that in future, when synthesizing up to the final Clifford, the calling application would like to obtain this Clifford as a part of the result (for instance, to modify the measurement operators that need to be performed), and we currently don't have API for any of this.